pytorch_lightning/loops/utilities.py

```diff
@@ -99,7 +99,7 @@ def _process_training_step_output(
     elif isinstance(training_step_output, torch.Tensor):
         loss = training_step_output
 
-    if trainer.terminate_on_nan:
+    if trainer.terminate_on_nan or trainer.detect_anomaly:
```

Comment:
```suggestion
    if trainer.terminate_on_nan and not trainer.detect_anomaly:
```
probably this? check with the test below (see my comment)